slides/E-async-web.md

```diff
@@ -0,0 +1,892 @@
+        handle_connection(socket).await?;
+    }
+}
+```
+
+---
+layout: default
+---
+# It works!
+
+```bash
+$ echo -e Ferris | nc localhost 6379
+Hello Ferris!⏎
+```
+
+*Question: But does it scale?*
+
+<v-click>
+<div>
+Nope! Only one request at a time!
+</div>
+</v-click>
+
+---
+layout: default
+---
+
+# Spawning tasks is cheap!
+```rust
+#[tokio::main]
+async fn main() -> Result<()> {
+    let listener = TcpListener::bind("127.0.0.1:6379").await.unwrap();
+
+    loop {
+        // The second item contains the IP and port of the new connection.
+        let (socket, _) = listener.accept().await.unwrap();
+        handle_connection(socket).await?;
+    }
+}
+```
+becomes:
+```rust
+async fn main() -> Result<()> {
+    loop {
+        // The second item contains the IP and port of the new connection.
+        let (socket, _) = listener.accept().await.unwrap();
+        tokio::task::spawn(async {
+            handle_connection(socket).await?;
+            Ok::<_, anyhow::Error>(()) // <-- cool trick to specify error type
+        });
+    }
+}
+```
+
+---
+layout: section
+---
+
+# Rust for web
+
+---
+layout: default
+---
+
+# [Are we web yet?](https://www.arewewebyet.org/)
+
+- "Yes! And it's freaking fast!"
+- Several web frameworks exist
+  - [`rocket`](https://rocket.rs/)
+  - [`actix-web`](https://actix.rs/)
+  - [`warp`](https://github.com/seanmonstar/warp)
+  - [`axum`](https://github.com/tokio-rs/axum)
+  - ...loads more
+- Several DB drivers and ORMs
+- Much more!
+
+---
+layout: default
+---
+
+# Axum demo: setting up server
+
+```rust
+use axum::{
+    extract::{Path, State},
+    response::Html,
+    routing::get,
+    Router,
+};
+use std::net::SocketAddr;
+
+#[tokio::main]
+async fn main() {
+    // set up shared, mutable state.
+    let app_state = Arc::new(Mutex::new(Vec::new()));
+    // build our application with a route
+    let app = Router::new()
+        .route("/:name", get(handler))
+        .with_state(app_state);
+    // run it
+    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+    println!("listening on {}", addr);
+    axum::Server::bind(&addr)
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+```
+
+---
+layout: default
+---
+# Axum demo: request hander
+
+```rust
+/// A very long type name warrants a type alias
+type AppState = State<Arc<Mutex<Vec<String>>>>;
+
+async fn handler(
+    Path(name): Path<String>,
+    State(past_names): State<AppState>,
+) -> Html<String> {
+    let mut response = format!("<h1>Hello, {name}!</h1>");
+
+    // Of course, locking here is not very fast
+    let mut past_names = past_names.lock().await;
+
+    if !past_names.is_empty() {
+        response += "<h2>Names we saw earlier:</h2>";
+        past_names
+            .iter()
+            .for_each(|name| response += &format!("<p>{name}</p>"))
+    }
+
+    past_names.push(name);
+
+    Html(response)
+}
+```
+
+---
+layout: section
+---
+# Bonus section: `Pin`
```

Comment:
I think I would advice being careful about this even if you have time; since `Pin` doesn't really do anything unless things are marked `!Unpin` (and the slides below don't introduce that concept)

If it is a really common problem in async code (I can't judge that--probably common once you start using references?), maybe a slightly higher level explanation can suffice + a pointer where students can read about it.